crates/xilem_masonry/src/sequence.rs

```diff
@@ -65,13 +68,75 @@ pub struct WasAView;
 /// See [`WasAView`]
 pub struct WasASequence;
 
-impl<State, Action, View: MasonryView<State, Action>> ViewSequence<State, Action, WasAView>
-    for View
+/// Two types are `SequenceCompatible` if a View with element type `SourceElement` can be used
+/// as the items of a sequence with element type `Self`.
+///
+/// For example, there may be a view trait which supports metadata with a sensible default
+/// around a "child" view type, e.g. a widget with optional layout metadata.
+///
+/// If the child type is sequence compatible with the type with metadata, the child type can
+/// be used with automatic conversion into this type with metadata.
+///
+/// This trait also supports for more advanced cases, including where the parent view is be an
+/// enum which is only sometimes the child view (with element `SourceElement`).
+/// `Self::from_mut` will panic when called with items which don't contain items of `SourceElement`
+pub trait SequenceCompatible<SourceElement: ViewElement>: ViewElement {
+    /// Convert from the source element into the type used in the sequence
+    fn into_item(element: SourceElement) -> Self;
+    /// Convert from the mutable form into the immutable form
+    ///
+    /// This may panic *if* the reference doesn't come from an element returned from `Self`
+    fn access_mut<R>(
+        reference: Self::Mut<'_>,
+        f: impl FnOnce(&mut SourceElement::Mut<'_>) -> R,
+    ) -> R;
+}
+
+impl<SourceElement: ViewElement, SequenceElement: ViewElement> SequenceCompatible<SourceElement>
```

Comment:
Maybe put all the bounds into the `where` clause (for ease of understanding this)? Or is this intended?